ejabberd: 20.03 -> 20.04

iblech opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

This is a simple version bump. "In addition to various bugfixes and improvements, this release massively improves audio and video calls support."

I have the new version deployed here and it works fine, but I just started hosting my own xmpp server yesterday so am not really qualified to judge.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

iblech committed:
ejabberd: 20.03 -> 20.04

ajs124 commented:
The test (in `nixos/tests/xmpp/ejabberd.nix`) doesn't seem to work anymore, but that's true on master with or without this PR.
So if you want to fix it, that would be really cool. We can probably merge this either way.

iblech commented:
Thank you for your feedback and for pointing me to the failing test (for any onlookers, [here is the log of the failing test](https://hydra.nixos.org/build/122152627/nixlog/21)). :-) I had a quick look. The test looks reasonable enough and I cannot really fathom why it should fail.

Perhaps it's just because ejabberd is too slow to create the `.erlang.cookie` file? If so, we should give it some more slack by adding directly after [line 253 of the test](https://github.com/NixOS/nixpkgs/blob/76605b21272f6bcc45a8d90009163c6f899024eb/nixos/tests/xmpp/ejabberd.nix#L253) the following instruction:

    server.sleep(15)

Unfortunately I cannot test this proposal, as I don't have access to any machine on which the flavor of qemu used by the test suite works.

ajs124 commented:
Ah, that's a different unrelated failure. Which might be fixed by a sleep in there, yes.
What I'm talking about is the client failing [like this](https://gist.github.com/ajs124/8ef26ee865cc7aaa9be4947d3744280e).

What's weird is that it works in the prosody test, which calls the exact same client code.

Aha! 8aea5288725688f7f71bf12c8ee1bb83147b22c6 is probably at fault then. The test works again when reverting it at least. @NinjaTrappeur could you maybe look into this? Seems like the ejabberd test might not be configured for some of the features this is trying to test now.

NinjaTrappeur commented:
Hey!

The test does not seems to enable `mod_disco` nor `mod_upload_http`. The issue is coming from there.

I'm having a look at it.

PS: Ofborg can now run the VM tests, you probably want to setup `passthru.tests` for this package to let the CI run those. It'll be simpler to see stuff breaking in the future. See https://github.com/NixOS/nixpkgs/blob/ec618a9120acb298e69afe9d02478722978781ec/pkgs/servers/xmpp/prosody/default.nix#L77

picnoir committed:
nixosTests.ejabberd: disabling mod_http_upload test

We're adding a flag to xmpp-sendmessage to disable the mod_http_upload
check to let ejabberd use this script with its current configuration.

The current Ejabberd test config does not set up a http_upload
endpoint. Configuring such a endpoint would probably require us to
create a proper TLS setup for the test.

This is not an ideal solution, but at least it'll let us run some
basic test on Ejabberd. Ideally, we should properly setup a
http_upload endpoint in the Ejabberd test to keep it on par with the
prosody one.

We're also pointing passthru.tests to the NixOS VM test to let the
Nixpkgs CI run the test on each Ejabberd update.

NinjaTrappeur commented:
After quite a bit of mud fighting with Ejabberd, I did not managed to setup `mod_upload_http` for this test. I suspect we'd need a proper TLS setup for the test. I'm not 100% sure about that though, the logs are pretty scarce. I guess somebody actually using Ejabbered would be a better candidate to fix that: I never used this software.

In the meantime, I wrote a small patch allowing us to disable the `http_upload` check for the Ejabbered test. I took advantage of this to add the passthru attribute I was talking about in the previous post.

I can't push anything to this branch, @iblech, you can apply the patch by running:

```
curl "https://github.com/NinjaTrappeur/nixpkgs/commit/deb1bcf3b221d2e626085b36d7da7e4647880448.patch" | git am
```

iblech commented:
Thanks to both of you for picking up this issue! :-) NinjaTrappeur, I applied your patch, thanks for the taking the effort and making it extra-convenient for me!

NinjaTrappeur commented:
You're welcome :)

Interesting, looks like ofborg is experiencing the same cookie issue you were seeing on hydra: https://logs.nix.ci/?key=nixos/nixpkgs.90462&attempt_id=4491813c-4b62-4904-8daf-4cf3ef7ec73e. That's weird, I do not experience that on my machine.

You can now run the test on ofborg by pushing to this branch (see the `ejabberd.passthru.tests` check), I'll let you find a workaround for that :)

iblech committed:
nixosTests.ejabberd: give ejabberd more time to start up

iblech commented:
Hm, very interesting!

First I just pushed `sleep(15)`, to see whether this actually makes the issue go away.

Then regarding a more robust solution: I was under the impression that `wait_for_unit` waits till the unit is in "active" state. But apparently ejabberd creates the cookie file only after systemd believes the unit to be active... Perhaps the correct fix would actually amend the systemd unit to create a placeholder cookie file before starting up ejabberd?

NinjaTrappeur commented:
Hey Ingo,

> Then regarding a more robust solution: I was under the impression that
> `wait_for_unit` waits till the unit is in "active" state. But
> apparently ejabberd creates the cookie file only after systemd
> believes the unit to be active...

Systemd will consider the unit active once the ExecStartPre script has
been executed and until the ExecStart script is running.

Without any further tweaking, systemd won't know anything about the
internal state of the application. IE. whether the service finished its
initialization procedure and is able to process incoming requests. The
Erlang VM-based apps are well known for having a pretty long boot time,
that's probably the issue we're hitting here.

> Perhaps the correct fix would actually amend the systemd unit to
> create a placeholder cookie file before starting up ejabberd?

I don't think this is the solution we're looking for. I think we should
rather make sure the Ejabberd is indeed responsive before quering it.

I see two solutions here.

1. We accept the fact the service might take some time to boot and fail
   to respond for a brief amount of time. In that case, we could replace
   the server.succeed here [1] with wait_until_succeeds.
2. We create a ejabberd-wait-online oneshot service that'll be triggered
   once the service gets online first. Something similar to [2]. You can
   then wait for this unit in the test. The question then is what's the
   condition to consider Ejabberd online. I can't really answer that :)

The solution 1 would be a nice way to fix the test. The solution 2 would
take that a step further and would create a new target you could
potentially use in some real world deployments; at the cost of being a
bit more costly to implement. I have no idea whether it worth the
trouble or not.

[1] https://github.com/NixOS/nixpkgs/blob/b20f9112d263e574672c004d6e2775d06ebb1fb7/nixos/tests/xmpp/ejabberd.nix#L255
[2] https://github.com/NixOS/nixpkgs/blob/b20f9112d263e574672c004d6e2775d06ebb1fb7/nixos/modules/system/boot/networkd.nix#L1207

ajs124 commented:
20.07 is out now + the fix in #97526 seems nicer than just disabling upload. The other parts of this PR, e.g. the `passthru.tests` and bump are still relevant, though.
